src-electron/upgrade/upgrade.js

```diff
@@ -0,0 +1,126 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+const dbApi = require('../db/db-api.js')
+const queryPackage = require('../db/query-package.js')
+const queryAttribute = require('../db/query-attribute.js')
+const queryCluster = require('../db/query-cluster.js')
+const dbEnum = require('../../src-shared/db-enum.js')
+const fs = require('fs')
+const fsp = fs.promises
+
+/**
+ * This file implements upgrade rules which are used to upgrade .zap files and xml files
+ * to be in sync with the spec
+ */
+
+/**
+ * Returns an array of objects containing global attributes that should be forced external.
+ *
+ * @export
+ * @param {*} db
+ * @param {*} attributeId
+ * @returns An array of objects
+ */
+
+async function getForcedExternalStorage(db, attributeId) {
+  let pkgs = await queryPackage.getPackageRefByAttributeId(db, attributeId)
+  let zcl = await queryPackage.getPackageByPackageId(db, pkgs)
+  zcl = zcl.path
+  let obj = await fsp.readFile(zcl)
+  let data = JSON.parse(obj)
+  let byName = data?.attributeAccessInterfaceAttributes
+  let lists = data?.listsUseAttributeAccessInterface
+  let forcedExternal = { byName, lists }
+  return forcedExternal
+}
+
+/**
+ * Returns a flag stating which type of storage option the attribute is categorized to be.
+ *
+ * @export
+ * @param {*} db
+ * @param {*} clusterName
+ * @param {*} clusterRef
+ * @param {*} storagePolicy
+ * @param {*} forcedExternal
+ * @param {*} attributeId
+ * @returns Storage Option
+ */
+
+async function computeStorageNewConfig(
+  db,
+  clusterRef,
+  storagePolicy,
+  forcedExternal,
+  attributeId
+) {
+  let storageOption
+  let attributeName
+  let clusterName
+  clusterName = await queryCluster.selectClusterName(db, clusterRef)
+  if (storagePolicy == dbEnum.storagePolicy.attributeAccessInterface) {
+    storageOption = dbEnum.storageOption.external
+  } else if (storagePolicy == dbEnum.storagePolicy.any) {
+    storageOption = dbEnum.storageOption.ram
+  } else {
+    throw 'check storage policy'
+  }
+  attributeName = await queryAttribute.selectAttributeName(db, attributeId)
+  if (
+    forcedExternal.byName &&
+    forcedExternal.byName[clusterName] &&
+    forcedExternal.byName[clusterName].includes(attributeName)
+  ) {
+    storageOption = dbEnum.storageOption.external
+  }
+  return storageOption
+}
+
+/**
+ * Returns a flag stating which type of storage option the attribute is categorized to be.
+ *
+ * @export
+ * @param {*} db
+ * @param {*} clusterName
+ * @param {*} forcedExternal
+ * @param {*} attributeId
+ * @returns Storage Policy
+ */
+
+async function computeStorageImport(
+  db,
+  clusterName,
+  forcedExternal,
+  attributeId
+) {
+  let storagePolicy
+  let attributeName
+  attributeName = await queryAttribute.selectAttributeName(db, attributeId)
```

Comment:
Pass the attribute name as an argument to this computeStorageImport function and get rid of this extra sql call to retrieve the name of the attribute.

Reply:
Ok will do